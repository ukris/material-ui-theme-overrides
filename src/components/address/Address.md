
Address Editable Version with Place Name
```jsx
    <Address
        placeName="Ontario, Canada"
        setShowEditor={() => true} updateCell={(placeId) => alert(`Place Name sent to Parent: ${placeId}`)} 
    />
```


Address Editable Version with no Place Name
```jsx
    <Address
        updateCell={(placeId) => alert(`Place Name sent to Parent: ${placeId}`)} 
    />
```