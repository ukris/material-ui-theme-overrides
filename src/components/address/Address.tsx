import React from 'react'
import parse from 'autosuggest-highlight/parse'
import throttle from 'lodash/throttle'
import ClickAwayListener from '@material-ui/core/ClickAwayListener'
import TextField from '@material-ui/core/TextField'
import Autocomplete from '@material-ui/lab/Autocomplete'
import LocationOnIcon from '@material-ui/icons/LocationOn'
import Grid from '@material-ui/core/Grid'
import Typography from '@material-ui/core/Typography'
import { makeStyles } from '@material-ui/core/styles'

import getGmapsEndpoint from 'endpoints/gmaps'
import mocksEnabled from 'endpoints/mocksEnabled'
import './styles.scss'

function loadScript(src: string, position: HTMLElement | null, id: string) {
  if (!position) {
    return
  }

  const script = document.createElement('script')
  script.setAttribute('async', '')
  script.setAttribute('id', id)
  script.src = src
  position.appendChild(script)
}

const autocompleteService = { current: null }

const useStyles = makeStyles((theme: any) => ({
  icon: {
    color: theme.palette.text.secondary,
    marginRight: theme.spacing(2),
  },
  input: {
    '& input': {
      fontSize: (theme && theme.fontSize ) ? theme.fontSize.md : '1rem'
    }
  }
}))

interface PlaceType {
  description: string
  structured_formatting: {
    main_text: string
    secondary_text: string
    main_text_matched_substrings: [
      {
        offset: number
        length: number
      },
    ]
  }
}


export interface EditorProps {
  /** Google Maps Place Name */
  placeName: string

  /** Signals if Place can be changed or not */
  editable?: boolean

  /** Updates row data in parent component with the purpose of persisting when switching table page */
  updateCell: (placeId: string) => void

  /** ClickAwayListener prop */
  setShowEditor?: (content:any) => void
}

export default function Address({ placeName, editable = true, updateCell, setShowEditor } : EditorProps) {
  const classes = useStyles()
  const [inputValue, setInputValue] = React.useState(placeName || 'Select Location')
  const [options, setOptions] = React.useState<PlaceType[]>([])
  const [selected, setSelected] = React.useState('')
  const [inputContainer, setInputContainer] = React.useState<HTMLElement | null>(null)
  const loaded = React.useRef(false)
  if (typeof window !== 'undefined' && !loaded.current) {
    if (!document.querySelector('#google-maps')) {
      loadScript(
        getGmapsEndpoint(inputValue),
        document.querySelector('head'),
        'google-maps',
      )
    }
    loaded.current = true
  }

  /* Method that gets the name of the new selected place and calls updateRow passed function */
  const handleNewSelectedPlace = (newSelectedVal: string) => {
    if(newSelectedVal !== selected) {
      const place = options.find((option: any) => option.description === newSelectedVal)
        // @ts-ignore
      if(place && place.description) { // backend data POST/PUT request should start inside this conditional
        setSelected(newSelectedVal)
        // @ts-ignore
        updateCell(place.description)
      }
    }
  }

  /* Method that obtains the name of the new selected place when user clicks item in list */
  const handleItemClick = (ev: any) => {
    const selectedItem = JSON.stringify(ev.target.parentElement.innerText).replace(/\\n\\n/g, ', ')
    handleNewSelectedPlace(selectedItem.replace(/"/g,''))
  }


  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    setInputValue(event.target.value)
  }

  const fetch = React.useMemo(
    () =>
      throttle((request: { input: string }, callback: (results?: PlaceType[]) => void) => {
        (autocompleteService.current as any).getPlacePredictions(request, callback)
      }, 200),
    [],
  )

  React.useEffect(() => {
    let active = true

    if (!autocompleteService.current && (window as any).google) {
      autocompleteService.current = new (window as any).google.maps.places.AutocompleteService()
    }
    if (!autocompleteService.current) {
      return undefined
    }

    if (inputValue === '') {
      setOptions([])
      return undefined
    }

    !mocksEnabled() ?
      fetch({ input: inputValue }, (results?: PlaceType[]) => {
        if (active) {
          setOptions(results || [])
        }
      })
      :
      import("mocks/gmaps").then(gmapsmock => {
        // @ts-ignore
        setOptions([...gmapsmock.places])
      })

    return () => {
      active = false
    }
  }, [inputValue, fetch])

  // Sets focus on input so user can start typing when component mounts
  if(inputValue === placeName && inputContainer !== null) {
    // @ts-ignore
    inputContainer.children[0].children[0].focus()
  }

  return (
    <div className="table-address" data-test="table-address">
      <ClickAwayListener onClickAway={()=> setShowEditor && setShowEditor(null)}>
          <Autocomplete
            id="table-address-autocomplete" // Required for snapshot testing
            open={true}
            getOptionLabel={(option) => (typeof option === 'string' ? option : option.description)}
            filterOptions={(x) => x}
            options={options}
            autoComplete
            includeInputInList
            disabled={!editable}
            // @ts-ignore
            defaultValue={inputValue}
            disableCloseOnSelect={true}
            renderInput={(params) => (
                <TextField
                {...params}
                ref={(inputContainer: HTMLElement | null) => setInputContainer(inputContainer)}
                label=""
                variant="outlined"
                fullWidth
                onChange={handleChange}
                className={classes.input}
                />)
            }
            renderOption={(option) => {
                const matches = option.structured_formatting.main_text_matched_substrings
                const parts = parse(
                option.structured_formatting.main_text,
                matches.map((match: any) => [match.offset, match.offset + match.length]),
                )
                return (
                <Grid container alignItems="center" onMouseUp={(ev) => handleItemClick(ev)}>
                    <Grid item>
                    <LocationOnIcon className={classes.icon} />
                    </Grid>
                    <Grid item xs>
                      {parts.map((part, index) => (
                        <span key={index} style={{ fontWeight: part.highlight ? 700 : 400 }}>
                        {part.text}
                        </span>
                    ))}
                    <Typography variant="body2" color="textSecondary">
                        {option.structured_formatting.secondary_text}
                    </Typography>
                    </Grid>
                </Grid>
                )
            }}
          />
      </ClickAwayListener>
    </div>
  )
}