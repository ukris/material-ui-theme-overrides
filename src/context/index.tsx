import CurrentBufferContext from './current-buffer-context'
import TableContext from './table-context'
import GlobalContext from './global-context'
import ProjectContext from './project-context'

export { CurrentBufferContext,  GlobalContext, ProjectContext, TableContext }