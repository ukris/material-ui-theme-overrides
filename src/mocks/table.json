{
  "id": 1,
  "editable": true,
  "people": [
      ["https://material-ui.com/static/images/avatar/1.jpg",
        "Remy Sharp"
      ],
      [
        "https://material-ui.com/static/images/avatar/2.jpg",
        "Jerko Raminoff"
      ],
      [
        "https://material-ui.com/static/images/avatar/3.jpg",
        "Cindy Baker"
      ],
      [
        "",
        "Dean Fremington"
      ],
      [
        "https://material-ui.com/static/images/avatar/5.jpg",
        "Winko Wiskardin"
      ],
      [
        "https://material-ui.com/static/images/avatar/4.jpg",
        "Paula Howard"
      ],
      [
        "",
        "Paula Escariante"
      ],
      [
        "https://material-ui.com/static/images/avatar/6.jpg",
        "Domino Ramp"
      ],
      [
        "https://material-ui.com/static/images/avatar/7.jpg",
        "Sarah Ilighis"
      ]
    ],
  "columns": [
      {
          "id": "id",
          "hidden": true
      },
      {
          "id":"name",
          "label":"DessertÂ (100g serving)",
          "type": "ellipsis",
          "width": 20
      },
      {
          "id":"bar",
          "label":"Progress",
          "type": "progressPercentage"
      },
      {
          "id":"calories",
          "label":"Calories",
          "type": "number",
          "width": 7
      },
      {
          "id":"assignments",
          "label":"Assigned to",
          "type": "people",
          "options": [
          ["https://material-ui.com/static/images/avatar/1.jpg",
              "Remy Sharp"
          ],
          [
              "https://material-ui.com/static/images/avatar/2.jpg",
              "Jerko Raminoff"
          ],
          [
              "https://material-ui.com/static/images/avatar/3.jpg",
              "Cindy Baker"
          ],
          [
              "",
              "Dean Fremington"
          ],
          [
              "https://material-ui.com/static/images/avatar/5.jpg",
              "Winko Wiskardin"
          ],
          [
              "https://material-ui.com/static/images/avatar/4.jpg",
              "Paula Howard"
          ],
          [
              "",
              "Paula Escariante"
          ],
          [
              "https://material-ui.com/static/images/avatar/6.jpg",
              "Domino Ramp"
          ],
          [
              "https://material-ui.com/static/images/avatar/7.jpg",
              "Sarah Ilighis"
          ]
          ]
      },
      {
          "id":"tags",
          "label":"Tags",
          "type": "tags"
      },
      {
        "id":"label",
        "label":"Label",
        "type": "label"
    },
      {
          "id":"date",
          "label":"Date",
          "type": "dateRange"
      },
      {
          "id":"rating",
          "label":"Rating",
          "type": "rating",
          "step": null,
          "options": [{
            "icon": "circle",
            "max": 5
          }]
      },
      { 
          "id": "location",
          "label": "Location",
          "type": "location"
      },
      { 
          "id": "flagWithISO",
          "label": "Country",
          "type": "flagWithISO"
      }
  ],
  "totalCount": 4,
  "rows":[
      [
        "1",
        "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque quis vestibulum augue, ut ullamcorper velit. Etiam sodales odio ut luctus efficitur",
        30,
        3.7,
        [
          ["https://material-ui.com/static/images/avatar/2.jpg","Jerko Raminoff"],
          ["","Dean Fremington"],
          ["https://material-ui.com/static/images/avatar/4.jpg","Paula Howard"]
        ],
        ["tag1","tag2","tag3","tag4"],
        "OK",
        [1596783600000, 1598166000000],
        1,
        "Ontario, Canada",
        "CA"
      ],  [
        "2",
        "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque quis vestibulum augue, ut ullamcorper velit. Etiam sodales odio ut luctus efficitur",
        30,
        3.7,
        [
          ["https://material-ui.com/static/images/avatar/2.jpg","Jerko Raminoff"],
          ["","Dean Fremington"],
          ["https://material-ui.com/static/images/avatar/4.jpg","Paula Howard"]
        ],
        ["tag1","tag2"],
        "OK1",
        [1596783600000, 1598166000000],
        1,
        "San Francisco, California, EE. UU.",
        "US"
      ],  [
        "1",
        "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque quis vestibulum augue, ut ullamcorper velit. Etiam sodales odio ut luctus efficitur",
        50,
        2.7,
        [
          ["https://material-ui.com/static/images/avatar/2.jpg","Jerko Raminoff"],
          ["","Dean Fremington"],
          ["https://material-ui.com/static/images/avatar/4.jpg","Paula Howard"]
        ],
        ["tag3","tag4"],
        "OK2",
        [1596783600000, 1598166000000],
        1,
        "Buenos Aires, CABA, Argentina",
        "AR"
      ],  [
        "1",
        "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque quis vestibulum augue, ut ullamcorper velit. Etiam sodales odio ut luctus efficitur",
        60,
        5,
        [
          ["https://material-ui.com/static/images/avatar/2.jpg","Jerko Raminoff"],
          ["","Dean Fremington"],
          ["https://material-ui.com/static/images/avatar/4.jpg","Paula Howard"]
        ],
        ["tag1","tag4"],
        "OK",
        [1596783600000, 1598166000000],
        1,
        "Santiago, Chile",
        "CL"
      ]
    ]
}