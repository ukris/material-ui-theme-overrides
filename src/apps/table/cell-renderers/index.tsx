import React, { useContext, useState, Suspense }from 'react'
import clsx from 'clsx'
import { format } from 'date-fns'

import TableCell from 'vendors/mui-table/TableCell'
import Star from '@material-ui/icons/Star'
import Circle from '@material-ui/icons/FiberManualRecord'
import MoodIcon from '@material-ui/icons/Mood'
import MoodBadIcon from '@material-ui/icons/MoodBad'
import AddIcon from '@material-ui/icons/Add'

import {RenderCountry} from 'components/country'
import { valToAvatars } from 'components/avatar'
import AvatarGroup from 'components/avatar-group/AvatarGroup'
import { Tag, Tags } from 'components/tags'
import { Label } from 'components/labels'
import ProgressBar from 'components/progressbar'
import { RenderCellType, CellValue, RenderCellProps } from '../types'
import { getWidth, roundFrac } from 'helpers'
import { GlobalContext, TableContext } from 'context'
import { tablePadding } from 'theme'
import useStyles from './style'

export  type HandleClick = (e: any) => void

export function renderLocation({classes, width, value, handleClick}: RenderCellProps) {
    const val:string = typeof value === 'string'? value : ''
    const { fontSize, ellipsis } = classes
    return (
        <div
            onClick={handleClick} title={val || 'Select a Place'}
            style={{width: getWidth(width)}}
            className={`${fontSize} ${ellipsis}`}
        >
            {val || ' '}
        </div>
    )    
}

export function renderPercentage({value, classes}: RenderCellProps) {
    const val = typeof value == 'number' ? value : 0
    const { widthCls } = classes
    return (
        <div className={clsx(widthCls )}>
            <ProgressBar value={val}/>
        </div>
    )
}

export function renderRating({classes, col, value, handleClick }: RenderCellProps) {
    const fontSizeLg = classes ? classes.fontSize : null
    const primaryColor = classes ? classes.primaryColor: null
    const val = typeof value == 'number' ? value : 0
    const { options } = col
    const option:any = options && options.length 
        ? options[0]
        : {
            max: 5, 
            icon: 'star'
        }
    const { icon = 'start', max = 5 }  = option
    const Icon = icon === 'star'
    ? <Star fontSize="inherit"/> 
    : 'circle' ? <Circle fontSize="inherit"/> 
    : 'mood' ? <MoodIcon fontSize="inherit"/> 
    : 'moodbad' ? <MoodBadIcon fontSize="inherit"/> :  <Star fontSize="inherit"/>
    const arr = new Array(max).fill(1)
    return ( 
        <div 
            className={clsx(fontSizeLg, 'inline')}
            onClick={(e)=>handleClick && handleClick(e)}
            >
           {arr.map((_,idx)=> {
                const cls = (idx+1) > val ? 'gray' : primaryColor
                const ico = (<span key={idx} className={clsx(cls)}>{Icon}</span>)
                return ico
            })}
        </div>
    )
}

export function  renderDateRange({classes, value, handleClick}: RenderCellProps) {
    let start:number = 0
    let end:number = 0
    const val:any = value as any
    if (Array.isArray(value) && value.length && value[0] && (!isNaN(val[0]))) {
        start = val[0]
        end = val[1] || 0
    }
    const formatStr = 'MMM dd'
    const fontSize = classes ? classes.fontSize : null
    const accent = classes ? classes.accent : null

    return (
        <Tag cls={clsx(fontSize, accent)} 
        onClick={handleClick}>
            {format(start, formatStr)}
            {' - '}
            {format(end, formatStr)}
        </Tag>
    )
}

export function  renderPeople({value, col, classes, handleClick}: RenderCellProps) {
    const AddI = handleClick ? AddIcon: null
    const avatars = valToAvatars(value as any[], col)

    const { paddingCls, widthCls } = classes
    return (<AvatarGroup
            avatars={avatars} 
            className={clsx(widthCls, paddingCls)}
            shape="circle" 
            size="sm" 
            max={3} 
            spacing="together" 
            Icon={AddI} 
            handleClick={handleClick}/>)
}

export function  renderImage({ value, handleClick}: RenderCellProps) {
    const val:any[] = Array.isArray(value) ? value: []
    const AddI = handleClick ? AddIcon: null
    return (<AvatarGroup avatars={val} shape="square" size="sm" max={3} spacing="separate" Icon={AddI}/>)
}

export function  renderISOWithFlag({value, handleClick}: RenderCellProps) {
    const val:string = typeof value === 'string'? value: ''
    return (<RenderCountry country={val} handleClick={handleClick}/>)
}

export function  renderTags({classes, col, value, width, handleClick}: RenderCellProps) {
    const fontSize = classes ? classes.fontSize : null
    const val:string[] = typeof value === 'string' ? [value] : Array.isArray(value)? value as string[]: []
    return(<Tags tags={val} 
        cls={clsx(fontSize,'inline')}
        css={{width: getWidth(width), minHeight: GlobalContext.fontSize.sm}}
        onClick={handleClick}/>
        )    
}
export function  renderLabel({classes, col, value, width, handleClick}: RenderCellProps) {
    const fontSize = classes ? classes.fontSize : null
    return(<Label value={value as string}
            onClick={handleClick}/>
        )    
}

export function renderEllipsis({col, value, handleClick, width, classes} : RenderCellProps) {
    const val:string = typeof value === 'string'? value: ''
    let cls:string = classes ? classes.fontSize : ''
    cls += ' truncate'

    return <div style={{width: getWidth(width)}}
                className={cls}
                onClick={(e)=> handleClick && handleClick(e)}>
                {val}
            </div>
}

const areEqual = (prevProps:RenderCellType, nextProps:RenderCellType) => {
    if (!prevProps.cell) return false
    return false
}

function  RenderCell(props:  RenderCellType) {
    const { cell } = props
    const { editable  } = useContext(TableContext)
    const [ newVal, setNewVal ] = useState<CellValue>(cell.value)
    const [showEditor, setShowEditor] = useState<any>(null)
    const classes = useStyles(props)
    //  @ts-ignore
    const { paddingCls, widthCls } = classes
    const col = GlobalContext.getColumn(cell.colIndex)
    const canEditThisCell = editable && !col.readOnly
   
    const { colIndex, rowIndex, rowId, value } = cell
    const id = `td-${rowIndex}-${colIndex}`
    const { align, disablePadding, renderer, width } = col
    const component = 'td'
    const scope = colIndex === 0 ? 'row' : undefined
    const adjWidth = width - tablePadding[0];

    let child: React.ReactElement | null = (typeof value !== 'object') ? <div className={clsx(widthCls, paddingCls )} style={{minWidth:getWidth(adjWidth)}}>{newVal}</div>: null
    const handleClick =  (e:any) => {
        e.stopPropagation()
        if (canEditThisCell) {
//      set only if different from current cell  setFocusedCell(cell)
            const { focusedCell }  = GlobalContext;
            if (!focusedCell || focusedCell.colIndex !== cell.colIndex || focusedCell.rowIndex !== cell.rowIndex) {
                let el = e.currentTarget.parentElement
                GlobalContext.setAnchorEl(el)
                GlobalContext.setFocusedCell(cell)
                const EditorCell = React.lazy(() => import('../cell-editors'))
                cell.value = newVal
                const editor = (
                    <Suspense fallback={null}>
                        <EditorCell 
                            cell={cell} 
                            setShowEditor={setShowEditor}
                            setNewVal={setNewVal}
                        />
                    </Suspense>
                )
                setShowEditor(editor)
            }
        }
    }
    if (renderer && !showEditor) {
        const props: RenderCellProps = {
            col,
            classes,
            value:newVal,
            width: adjWidth
        }
        props.handleClick = handleClick
        const newChild = renderer(props)
        if (newChild) {
            child = newChild
        }
    }
    const dataCol = `${colIndex}`
    const dataRow = `${rowIndex}|${rowId}`
    const cWidth = getWidth(col.width)
    return (
        <TableCell
            id={id}
            key={colIndex}
            align={align}
            component={component} 
            scope={scope}
            padding={disablePadding ? 'none' : 'default'}
            dataCol={dataCol}
            dataRow={dataRow}
            onClick={(ev:any) => handleClick && handleClick(ev)}
            style={{minWidth: cWidth, maxWidth: cWidth, marginRight: `${tablePadding[0]}rem`}}
        >
            {showEditor ? showEditor : child}
        </TableCell>
    )
}

export default React.memo(RenderCell, areEqual)