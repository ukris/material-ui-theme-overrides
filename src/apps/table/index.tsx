import React, { useContext, useEffect, useState } from 'react'
import {Table , TableBody, TableCell, TableContainer, TableRow} from '@material-ui/core'
import Pagination from '@material-ui/lab/Pagination'
import Paper from '@material-ui/core/Paper'
import Checkbox from '@material-ui/core/Checkbox'
import FormControlLabel from '@material-ui/core/FormControlLabel'
import Switch from '@material-ui/core/Switch'
// search - https://github.com/uqix/reactkit-table
// https://github.com/techniq/mui-virtualized-table
// https://github.com/bnabriss/react-mui-pagination
import DragIndicator from '@material-ui/icons/DragIndicator'

import TableHead from './table-head'
import TableToolbar from './table-toolbar'
import RenderCell from './cell-renderers'
import { Cell, TableType } from './types'
import { KeyAction } from 'helpers/table/keycode-mouse-events'

import { getLookupId, getSorting, sortRows, } from 'helpers/table'
import { FONT_BASIS } from 'theme'
import { TableContext, GlobalContext } from 'context'
import useStyles from './styles'

function EnTable(props: TableType) {
  const { rowsData, defaultOrderBy = 'name', desc = false, editable=false } = props

  const {
    columns,
    setColumns,
    rows, 
    setRows,
    domainId=-1,
    projectId=-1,
    boardId = -1,
    laneId = ''
  } = GlobalContext
  const {
    setEditable
  } = useContext(TableContext)
  const lookupId = getLookupId({domainId, projectId, boardId, laneId})
  const classes = useStyles(props)
  const [order, setOrder] = useState(desc)
  const [orderBy, setOrderBy] = useState(defaultOrderBy)
  const [selected, setSelected] = useState<any[]>([])
  const [page, setPage] = useState(0)
  const [dense, setDense] = useState(false)
  const [rowsPerPage, setRowsPerPage] = useState(5)
  const [collapsed, setCollapased] = useState(false)
  const [isLoading, setIsLoading] = useState(true)
  const [focusedCell, setFocusedCell] = useState<Cell|null>(null)
  const [] = useState<KeyAction|null>(null)
  const colDefs = props.columns

  useEffect(() => {
    setRows(rowsData)
    setColumns(colDefs)
    setIsLoading(false)
    setEditable(editable)
  },[rowsData, setRows, editable, setEditable, colDefs, setColumns])
  
  if (isLoading) {
    return null
  }
  const handleRequestSort = (event: any, property: any) => {
    const isDesc = orderBy === property && order
    setOrder(isDesc)
    setOrderBy(property)
  }
  
  const handleSelectAllClick = (event: any) => {
    if (event.target.checked) {
      const newSelecteds = rows.map((row,index) => index)
      setSelected(newSelecteds)
      return
    }
    setSelected([])
  }

  const handleClick = (event: any, index: number) => {
    const selectedIndex = selected.indexOf(index)
    let newSelected: any[] = []

    if (selectedIndex === -1) {
      newSelected = newSelected.concat(selected, index)
    } else if (selectedIndex === 0) {
      newSelected = newSelected.concat(selected.slice(1))
    } else if (selectedIndex === selected.length - 1) {
      newSelected = newSelected.concat(selected.slice(0, -1))
    } else if (selectedIndex > 0) {
      newSelected = newSelected.concat(
        selected.slice(0, selectedIndex),
        selected.slice(selectedIndex + 1),
      )
    }
    setSelected(newSelected)
  }

  const handleChangePage = (event: any, newPage: any) => {
    setPage(newPage)
  }

  const handleChangeDense = (event: any) => {
    setDense(event.target.checked)
  }
  // @ts-ignore
  const isSelected = (index: number) => selected.indexOf(index) !== -1

  const emptyRows = rowsPerPage - Math.min(rowsPerPage, rows.length - page * rowsPerPage)
  const id = 'enhanced-table'

  return (
    <div className={classes.root} data-test="table-container">
      <Paper className={classes.paper} elevation={0}>
        <TableToolbar numSelected={selected.length}
          lookupId={lookupId}
          collapsed={collapsed}
          setCollapsed={setCollapased}
          editable={editable}
        />
        <TableContainer>
          <Table
            className={classes.table}
            aria-labelledby="tableTitle"
            size={dense ? 'small' : 'medium'}
            aria-label="table"
            id = {id}
          >
            <TableHead
              numSelected={selected.length}
              desc={order}
              orderBy={orderBy}
              onSelectAll={handleSelectAllClick}
              onSort={handleRequestSort}
              rowCount={rows.length}
              columns={columns}
            />
            <TableBody>
              {sortRows(rows, getSorting(order, orderBy))
                .slice(page * rowsPerPage, page * rowsPerPage + rowsPerPage)
                .map((row: any, index: number) => {
                  const isItemSelected = isSelected(index)
                  const rowId = `${id}-row-${index}`
                  return (
                    <TableRow
                      hover
                      onClick={(event:any) => handleClick(event, index)}
                      role="checkbox"
                      aria-checked={isItemSelected}
                      tabIndex={-1}
                      key={index}
                      selected={isItemSelected}
                      id={rowId}
                    >
                      <TableCell style={{width:'5px'}} padding="none" 
                        id={`${index}|${row.id}`}>
                        <DragIndicator />
                      </TableCell>
                      <TableCell 
                        padding="none"
                        id={`${index}|${row.id}`}>
                        <Checkbox
                          color="primary"
                          checked={isItemSelected}
                          inputProps={{ 'aria-labelledby': rowId }}
                        />
                      </TableCell>
                      {columns.map((col, idx) => {
                        if (col.hidden || col.hide) {
                          return null
                        }
                        const thisCell = {rowId: row.id, rowIndex: index, colIndex: idx, value:row[idx]}
                        return (
                          <RenderCell
                            key={`${index}-${idx}`}
                            cell={thisCell}
                          />
                        )
                      })}
                    </TableRow>
                  )
                })}
              {emptyRows > 0 && (
                <TableRow style={{ height: (dense ? 12 * FONT_BASIS : 15 * FONT_BASIS) * emptyRows }}>
                  <TableCell 
                    colSpan={6}
                    />
                </TableRow>
              )}
            </TableBody>
          </Table>
        </TableContainer>
        <Pagination 
          page={page + 1} 
          onChange={handleChangePage}
          count={5}
          color="primary"
        />
      </Paper>
      <FormControlLabel
        control={<Switch checked={dense} onChange={handleChangeDense} />}
        label="Dense padding"
      />
    </div>
  )
}

export default React.memo(EnTable)