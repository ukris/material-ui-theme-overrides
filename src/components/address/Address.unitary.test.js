import React from 'react'
import renderer from 'react-test-renderer'
import { mount, configure } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import { GlobalContext } from 'context'
import { ThemeContext, ThemeProvider } from 'theme'
import { renderLocation } from '../table/cell-renderers'
import Address from './Address'

configure({ adapter: new Adapter() })

const themeZoom = GlobalContext.themeZoom;
const setThemeZoom = () => true
 
const value = "Ontario, Canada"

describe('Address | Read-mode', () => {
    const classes = { fontSize: 'makeStyles-fontSize-698', ellipsis: 'makeStyles-ellipsis-709'}
    const handleClick = () => false

    it('renders component with place', () => {
        expect(renderLocation({classes, value, handleClick})).toEqual(
            <span className="makeStyles-fontSize-698 makeStyles-ellipsis-709" onClick={handleClick} title="Ontario, Canada">Ontario, Canada</span>
        )
    });

    it('renders component with no place', () => {
        expect(renderLocation({classes, handleClick})).toEqual(
            <span className="makeStyles-fontSize-698 makeStyles-ellipsis-709" onClick={handleClick} title="Select a Place">{' '}</span>
        )
    });
})

describe('Address | Editable | No default place', () => {
    global.document.createRange = () => ({
        setStart: () => {},
        setEnd: () => {},
        commonAncestorContainer: {
          nodeName: "BODY",
          ownerDocument: document,
        },
    })

    const component = (
      <ThemeContext.Provider value={{ themeZoom, setThemeZoom }}>
        <ThemeProvider>
            <Address setShowEditor={() => true} updateCell={(placeId) => alert(`Place Name sent to Parent: ${placeId}`)}  />
        </ThemeProvider>
      </ThemeContext.Provider>
    )
    const wrapper = mount(component)
    const input = wrapper.find('input')

    it('renders component correctly', () => {
        const divContainer = wrapper.find('[data-test="table-address"]')

        expect(divContainer.length).toEqual(1)
        expect(input.length).toEqual(1)
        expect(input.props().value).toEqual('Select Location')
    });

    it('matches snapshot', () => {
        const tree = renderer.create(component,).toJSON()

        expect(tree).toMatchSnapshot();
    });
})

describe('Address | Editable | Default place', () => {
    global.document.createRange = () => ({
        setStart: () => {},
        setEnd: () => {},
        commonAncestorContainer: {
          nodeName: "BODY",
          ownerDocument: document,
        },
    })

    const component = (
      <ThemeContext.Provider value={{ themeZoom, setThemeZoom }}>
        <ThemeProvider>
            <Address placeName={value} setShowEditor={() => true} updateCell={(placeId) => alert(`Place Name sent to Parent: ${placeId}`)}  />
        </ThemeProvider>
      </ThemeContext.Provider>
    )
    const wrapper = mount(component)
    const input = wrapper.find('input')

    it('renders component correctly', () => {
        const divContainer = wrapper.find('[data-test="table-address"]')

        expect(divContainer.length).toEqual(1)
        expect(input.length).toEqual(1)
        expect(input.props().value).toEqual('Ontario, Canada')
    });
})
