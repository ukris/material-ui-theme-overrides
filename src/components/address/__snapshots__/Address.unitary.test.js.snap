// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Address | Editable | No default place matches snapshot 1`] = `
Array [
  " ",
  <div
    className="table-address"
    data-test="table-address"
  >
    <div
      aria-expanded={true}
      aria-owns="table-address-autocomplete-popup"
      className="MuiAutocomplete-root MuiAutocomplete-hasClearIcon MuiAutocomplete-hasPopupIcon"
      onClick={[Function]}
      onKeyDown={[Function]}
      onMouseDown={[Function]}
      onTouchEnd={[Function]}
      role="combobox"
    >
      <div
        className="MuiFormControl-root MuiTextField-root makeStyles-input-236 MuiFormControl-fullWidth"
      >
        
        <div
          className="MuiInputBase-root MuiOutlinedInput-root MuiAutocomplete-inputRoot MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiOutlinedInput-adornedEnd"
          onClick={[Function]}
        >
          <input
            aria-activedescendant=""
            aria-autocomplete="both"
            aria-controls="table-address-autocomplete-popup"
            aria-invalid={false}
            autoCapitalize="none"
            autoComplete="off"
            autoFocus={false}
            className="MuiInputBase-input MuiOutlinedInput-input MuiAutocomplete-input MuiAutocomplete-inputFocused MuiInputBase-inputAdornedEnd MuiOutlinedInput-inputAdornedEnd"
            disabled={false}
            id="table-address-autocomplete"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onMouseDown={[Function]}
            required={false}
            spellCheck="false"
            type="text"
            value=""
          />
          <div
            className="MuiAutocomplete-endAdornment"
          >
            <button
              aria-label="Clear"
              className="MuiButtonBase-root MuiIconButton-root MuiAutocomplete-clearIndicator MuiAutocomplete-clearIndicatorDirty"
              disabled={false}
              onBlur={[Function]}
              onClick={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={-1}
              title="Clear"
              type="button"
            >
              <span
                className="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  className="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                  />
                </svg>
              </span>
            </button>
            <button
              aria-label="Close"
              className="MuiButtonBase-root MuiIconButton-root MuiAutocomplete-popupIndicator MuiAutocomplete-popupIndicatorOpen"
              disabled={false}
              onBlur={[Function]}
              onClick={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={-1}
              title="Close"
              type="button"
            >
              <span
                className="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  className="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M7 10l5 5 5-5z"
                  />
                </svg>
              </span>
            </button>
          </div>
          <fieldset
            aria-hidden={true}
            className="PrivateNotchedOutline-root-320 MuiOutlinedInput-notchedOutline"
            style={
              Object {
                "paddingLeft": 8,
              }
            }
          >
            <legend
              className="PrivateNotchedOutline-legend-321"
              style={
                Object {
                  "width": 0.01,
                }
              }
            >
              <span
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "&#8203;",
                  }
                }
              />
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
  </div>,
]
`;
